include/mbedtls/rsa.h

```diff
@@ -970,6 +970,72 @@ int mbedtls_rsa_rsassa_pkcs1_v15_sign( mbedtls_rsa_context *ctx,
                                const unsigned char *hash,
                                unsigned char *sig );
 
+/**
+ * \brief          This function performs a PKCS#1 v2.1 PSS signature
+ *                 operation (RSASSA-PSS-SIGN).
+ *
+ * \note           The \p hash_id in the RSA context is the one used for the
+ *                 encoding. \p md_alg in the function call is the type of hash
+ *                 that is encoded. According to <em>RFC-3447: Public-Key
+ *                 Cryptography Standards (PKCS) #1 v2.1: RSA Cryptography
+ *                 Specifications</em> it is advised to keep both hashes the
+ *                 same.
+ *
+ * \note           This function always uses the maximum possible salt size,
+ *                 up to the length of the payload hash. This choice of salt
+ *                 size complies with FIPS 186-4 §5.5 (e) and RFC 8017 (PKCS#1
+ *                 v2.2) §9.1.1 step 3. Furthermore this function enforces a
+ *                 minimum salt size which is the hash size minus 2 bytes. If
+ *                 this minimum size is too large given the key size (the salt
+ *                 size, plus the hash size, plus 2 bytes must be no more than
+ *                 the key size in bytes), this function returns
+ *                 #MBEDTLS_ERR_RSA_BAD_INPUT_DATA.
+ *
+ * \deprecated     It is deprecated and discouraged to call this function
```

Comment:
Since this is a new function, I think it shouldn't have a `mode` argument.

Reply:
Agreed, if it's deprecated, it makes sense to remove it for a new function.

Reply:
@gilles-peskine-arm : Working on it, I just realized that if we remove the `mode` argument from the `_ext()` function, we cannot simply have the original function call the `_ext()` function without changing the original function's behavior: it will not support `MBEDTLS_RSA_PUBLIC` anymore.

As I see it, we have several options:
1. leave  the `mode` argument in the `_ext()` function.
2. remove the `mode` argument from the `_ext()` function and either:
    1. leave the original function intact and have the `_ext()` be a completely separate copy.
    2. accept the fact that `MBEDTLS_RSA_PUBLIC` is not supported anymore in the original function and have it return `MBEDTLS_ERR_RSA_BAD_INPUT_DATA` in this case. But this looks like a **breaking change** to me. I don't know enough to judge whether this would be acceptable or not.

What would you suggest?

Reply:
I would prefer 2.iii: change the existing function to `static int rsa_rsassa_pss_sign` taking both `mode` and `saltlen`, and make `mbedtls_rsa_rsassa_pss_sign` and `mbedtls_rsa_rsassa_pss_sign_ext` trivial wrappers around that.

We'll remove the ability to use the wrong `mode` in Mbed TLS 3, but as much as I think it needs to go, it needs to stay for a few more months.

Reply:
Deuh, sorry for the brainfart, I should have thought of that :-/